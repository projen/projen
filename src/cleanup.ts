import * as fs from "fs";
import * as path from "path";
import * as glob from "fast-glob";
import { PROJEN_DIR, PROJEN_MARKER } from "./common";
import * as logging from "./logging";

export const FILE_MANIFEST = `${PROJEN_DIR}/files.json`;

export function cleanup(dir: string, newFiles: string[], exclude: string[]) {
  try {
    const manifestFiles = getFilesFromManifest(dir);
    if (manifestFiles.length > 0) {
      // Use `FILE_MANIFEST` to remove files that are no longer managed by projen
      removeFiles(findOrphanedFiles(dir, manifestFiles, newFiles));
    } else {
      // Remove all files managed by projen with legacy logic
      removeFiles(findGeneratedFiles(dir, exclude));
    }
  } catch (e: any) {
    logging.warn(`warning: failed to clean up generated files: ${e.stack}`);
  }
}

function removeFiles(files: string[]) {
  for (const file of files) {
    fs.rmSync(file, { force: true, recursive: true });
  }
}

function findGeneratedFiles(dir: string, exclude: string[]) {
  const ignore = [
    ...readGitIgnore(dir),
    "node_modules/**",
    ...exclude,
    ".git/**",
  ];

  const files = glob.sync("**", {
    ignore,
    cwd: dir,
    dot: true,
    onlyFiles: true,
    followSymbolicLinks: false,
    absolute: true,
  });

  const generated = new Array<string>();

  for (const file of files) {
    const contents = fs.readFileSync(file, "utf-8");

    if (contents.includes(PROJEN_MARKER)) {
      generated.push(file);
    }
  }

  return generated;
}

function findOrphanedFiles(
  dir: string,
  oldFiles: string[],
  newFiles: string[]
) {
  return oldFiles
    .filter((old) => !newFiles.includes(old))
    .map((f: string) => path.resolve(dir, f));
}

function getFilesFromManifest(dir: string): string[] {
  try {
    const fileManifestPath = path.resolve(dir, FILE_MANIFEST);
    if (fs.existsSync(fileManifestPath)) {
      const fileManifest = JSON.parse(
        fs.readFileSync(fileManifestPath, "utf-8")
      );
      if (fileManifest.files) {
        return fileManifest.files;
      }
    }
  } catch (e: any) {
    logging.warn(
      `warning: unable to get files to clean from file manifest: ${e.stack}`
    );
  }

  return [];
}

function readGitIgnore(dir: string) {
  const filepath = path.join(dir, ".gitignore");
  if (!fs.existsSync(filepath)) {
    return [];
  }

  return fs
    .readFileSync(filepath, "utf-8")
    .split("\n")
    .filter((x) => x?.trim() !== "")
    .filter((x) => !x.startsWith("#") && !x.startsWith("!"))
    .map((x) => x.replace(/^\//, "")) // remove "/" prefix
    .map((x) => `${x}\n${x}/**`)
    .join("\n")
    .split("\n");
}
