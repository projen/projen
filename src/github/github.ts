import { GitHubActionsProvider } from "./actions-provider";
import { Dependabot, DependabotOptions } from "./dependabot";
import { GithubCredentials } from "./github-credentials";
import { MergeQueue, MergeQueueOptions } from "./merge-queue";
import { Mergify, MergifyOptions } from "./mergify";
import { PullRequestTemplate } from "./pr-template";
import {
  PullRequestBackport,
  PullRequestBackportOptions,
} from "./pull-request-backport";
import { PullRequestLint, PullRequestLintOptions } from "./pull-request-lint";
import { GithubWorkflow } from "./workflows";
import { Component } from "../component";
import { Project } from "../project";

export interface GitHubOptions {
  /**
   * Whether mergify should be enabled on this repository or not.
   *
   * @default true
   */
  readonly mergify?: boolean;

  /**
   * Options for Mergify.
   *
   * @default - default options
   */
  readonly mergifyOptions?: MergifyOptions;

  /**
   * Whether a merge queue should be used on this repository to merge pull requests.
   * Requires additional configuration of the repositories branch protection rules.
   *
   * @default false
   */
  readonly mergeQueue?: boolean;

  /**
   * Options for MergeQueue.
   *
   * @default - default options
   */
  readonly mergeQueueOptions?: MergeQueueOptions;

  /**
   * Enables GitHub workflows. If this is set to `false`, workflows will not be created.
   *
   * @default true
   */
  readonly workflows?: boolean;

  /**
   * Add a workflow that allows backport of PRs to other branches using labels.
   * When opening a new PR add a backport label to it,
   * and the PR will be backported to the target branches once the PR is merged.
   *
   * Should not be used together with mergify.
   *
   * @default false
   */
  readonly pullRequestBackport?: boolean;

  /**
   * Options for configuring pull request backport.
   *
   * @default - see defaults in `PullRequestBackportOptions`
   */
  readonly pullRequestBackportOptions?: PullRequestBackportOptions;

  /**
   * Add a workflow that performs basic checks for pull requests, like
   * validating that PRs follow Conventional Commits.
   *
   * @default true
   */
  readonly pullRequestLint?: boolean;

  /**
   * Options for configuring a pull request linter.
   *
   * @default - see defaults in `PullRequestLintOptions`
   */
  readonly pullRequestLintOptions?: PullRequestLintOptions;

  /**
   * Choose a method of providing GitHub API access for projen workflows.
   *
   * @default - use a personal access token named PROJEN_GITHUB_TOKEN
   */
  readonly projenCredentials?: GithubCredentials;

  /**
   * The name of a secret which includes a GitHub Personal Access Token to be
   * used by projen workflows. This token needs to have the `repo`, `workflows`
   * and `packages` scope.
   *
   * @default "PROJEN_GITHUB_TOKEN"
   * @deprecated - use `projenCredentials`
   */
  readonly projenTokenSecret?: string;

  /**
   * Download files in LFS in workflows
   *
   * @default true if the associated project has `lfsPatterns`, `false` otherwise
   */
  readonly downloadLfs?: boolean;
}

export class GitHub extends Component {
  /**
   * Returns the `GitHub` component of a project or `undefined` if the project
   * does not have a GitHub component.
   */
  public static of(project: Project): GitHub | undefined {
    const isGitHub = (c: Component): c is GitHub => c instanceof GitHub;
    return project.components.find(isGitHub);
  }

  /**
   * The `Mergify` component configured on this repository
   * This is `undefined` if Mergify is not enabled for this repository.
   */
  public readonly mergify?: Mergify;

  /**
   * The `MergeQueue` component configured on this repository
   * This is `undefined` if merge queues are not enabled for this repository.
   */
  public readonly mergeQueue?: MergeQueue;

  /**
   * Are workflows enabled?
   */
  public readonly workflowsEnabled: boolean;

  /**
   * GitHub API authentication method used by projen workflows.
   */
  public readonly projenCredentials: GithubCredentials;

  /**
   * The GitHub Actions provider used to manage the versions of actions used in steps
   */
  public readonly actions: GitHubActionsProvider;

  private readonly _downloadLfs?: boolean;

  public constructor(project: Project, options: GitHubOptions = {}) {
    super(project);

    this.actions = new GitHubActionsProvider();

    this.workflowsEnabled = options.workflows ?? true;

    this._downloadLfs = options.downloadLfs;

    if (options.projenCredentials && options.projenTokenSecret) {
      throw new Error(
        "projenTokenSecret is deprecated, please use projenCredentials instead"
      );
    }

    if (options.projenTokenSecret) {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: options.projenTokenSecret,
      });
    } else if (options.projenCredentials) {
      this.projenCredentials = options.projenCredentials;
    } else {
      this.projenCredentials = GithubCredentials.fromPersonalAccessToken({
        secret: "PROJEN_GITHUB_TOKEN",
      });
    }

    if (options.mergify ?? true) {
      this.mergify = new Mergify(this, options.mergifyOptions);
    }

    if (options.mergeQueue ?? false) {
      this.mergeQueue = new MergeQueue(this, options.mergeQueueOptions);
    }

    if (options.pullRequestLint ?? true) {
      new PullRequestLint(this, options.pullRequestLintOptions);
    }

    if (options.pullRequestBackport ?? false) {
      if (options.mergify) {
        this.project.logger.warn(
          "pullRequestBackport should not be used with mergify as mergify provides its own backport functionality. Please disable pullRequestBackport or mergify."
        );
      }
      new PullRequestBackport(this, options.pullRequestBackportOptions);
    }
  }

  /**
   * All workflows.
   */
  public get workflows(): GithubWorkflow[] {
    const isWorkflow = (c: Component): c is GithubWorkflow =>
      c instanceof GithubWorkflow;
    return this.project.components
      .filter(isWorkflow)
      .sort((w1, w2) => w1.name.localeCompare(w2.name));
  }

  /**
   * Adds a workflow to the project.
   * @param name Name of the workflow
   * @returns a GithubWorkflow instance
   */
  public addWorkflow(name: string) {
    const workflow = new GithubWorkflow(this, name);
    return workflow;
  }

  public addPullRequestTemplate(...content: string[]) {
    return new PullRequestTemplate(this, { lines: content });
  }

  public addDependabot(options?: DependabotOptions) {
    return new Dependabot(this, options);
  }

  /**
   * Finds a GitHub workflow by name. Returns `undefined` if the workflow cannot be found.
   * @param name The name of the GitHub workflow
   */
  public tryFindWorkflow(name: string): undefined | GithubWorkflow {
    return this.workflows.find((w) => w.name === name);
  }

  /**
   * Whether downloading from LFS is enabled for this GitHub project
   */
  public get downloadLfs() {
    return this._downloadLfs ?? this.project.gitattributes.hasLfsPatterns;
  }
}
