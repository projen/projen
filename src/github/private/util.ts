export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
