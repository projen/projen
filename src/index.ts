export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
