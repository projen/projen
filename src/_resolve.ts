import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
