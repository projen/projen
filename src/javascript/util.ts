import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
