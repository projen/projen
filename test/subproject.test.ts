import * as fs from "fs";
import * as path from "path";
import { synthSnapshot, TestProject } from "./util";
import { Project, TextFile, ProjectOptions, JsonFile, Component } from "../src";
import { PROJEN_MARKER } from "../src/common";
import { GitHubProject } from "../src/github";

test("composing projects declaratively", () => {
  const comp = new TestProject();
  new Project({
    name: "foo",
    parent: comp,
    outdir: path.join("packages", "foo"),
  });
  comp.synth();

  // THEN
  expect(
    fs.existsSync(path.join(comp.outdir, "packages", "foo", ".gitignore"))
  ).toBeTruthy();
});

test("composing projects synthesizes to subdirs", () => {
  // GIVEN
  const comp = new TestProject();

  // WHEN
  new Project({
    name: "foo",
    parent: comp,
    outdir: path.join("packages", "foo"),
  });
  new Project({
    name: "bar",
    parent: comp,
    outdir: path.join("packages", "bar"),
  });

  comp.synth();

  // THEN
  expect(fs.existsSync(path.join(comp.outdir, "README.md")));
  expect(
    fs.existsSync(path.join(comp.outdir, "packages", "foo", ".gitignore"))
  );
  expect(
    fs.existsSync(path.join(comp.outdir, "packages", "bar", ".gitignore"))
  );
});

test("errors when paths overlap", () => {
  // GIVEN
  const comp = new TestProject();
  new Project({
    name: "foo",
    parent: comp,
    outdir: path.join("packages", "foo"),
  });

  // WHEN/THEN
  expect(
    () =>
      new Project({
        name: "foo",
        parent: comp,
        outdir: path.join("packages", "foo"),
      })
  ).toThrow(/There is already a subproject with/i);
});

test("multiple levels", () => {
  const root = new TestProject();
  const child1 = new Project({
    name: "child1",
    parent: root,
    outdir: "child1",
  });
  const child2 = new Project({
    name: "child2",
    parent: child1,
    outdir: "child2",
  });

  expect(child1.outdir).toEqual(path.join(root.outdir, "child1"));
  expect(child2.outdir).toEqual(path.join(root.outdir, "child1", "child2"));
});

test("subprojects cannot introduce files that override each other", () => {
  const root = new TestProject();
  const child = new Project({
    name: "subproject",
    parent: root,
    outdir: "subproject",
  });

  new TextFile(root, "subproject/file.txt");
  expect(() => new TextFile(child, "file.txt")).toThrow(
    /There is already a file under subproject(\\|\/)file\.txt/
  );
});

test('"outdir" for subprojects must be relative', () => {
  const root = new TestProject();
  expect(
    () => new Project({ name: "foobar", parent: root, outdir: "/foo/bar" })
  ).toThrow(/"outdir" must be a relative path/);
});

test("subproject generated files do not get cleaned up by parent project", () => {
  const root = new TestProject();
  const child = new PreSynthProject({ parent: root, outdir: "subproject" });

  // no files have been generated yet
  expect(fs.existsSync(child.file.absolutePath)).toEqual(false);

  // generate all project files at least once
  root.synth();
  expect(child.fileExistedDuringPresynth).toEqual(false);
  expect(fs.existsSync(child.file.absolutePath)).toEqual(true);

  // resynthesize projects with all generated files already existing
  root.synth();
  expect(child.fileExistedDuringPresynth).toEqual(true);
  expect(fs.existsSync(child.file.absolutePath)).toEqual(true);
});

test("subproject generated json files can be synthed", () => {
  const root = new TestProject();
  const child = new PreSynthProject({ parent: root, outdir: "subproject" });
  new JsonFile(child, "test.jsonc", {
    marker: true,
    allowComments: true,
    obj: {
      test: "data",
    },
  });

  const out = synthSnapshot(root);
  expect(out["subproject/test.jsonc"]).toMatchInlineSnapshot(`
    {
      "test": "data",
    }
  `);
});

test("subprojects do not add a Projenrc component", () => {
  // GIVEN
  const parent = new TestProject();

  const child = new TestProject({
    parent,
    outdir: "sub",
    projenrcJson: true,
  });

  // THEN
  const rcFiles = child.components.filter((o: Component) =>
    o.constructor.name.toLowerCase().includes("projenrc")
  );
  expect(rcFiles.length).toBe(0);
});

test("subprojects use root level default task", () => {
  // GIVEN
  const root = new TestProject();

  const childOne = new TestProject({
    parent: root,
    outdir: "one",
  });

  new TestProject({
    parent: childOne,
    outdir: "two",
  });

  // THEN
  const out = synthSnapshot(root);
  expect(out["one/.projen/tasks.json"]).toMatchSnapshot();
  expect(out["one/.projen/tasks.json"].tasks.default.steps).toEqual([
    {
      cwd: "..",
      exec: "npx projen default",
    },
  ]);
  expect(out["one/two/.projen/tasks.json"]).toMatchSnapshot();
  expect(out["one/two/.projen/tasks.json"].tasks.default.steps).toEqual([
    {
      cwd: "../..",
      exec: "npx projen default",
    },
  ]);
});

test("files are annotated as generated on subproject", () => {
  // GIVEN

  // Currently `Project` has an empty implementation of `annotateGenerated`,
  // so we use `GitHubProject` to observe the resulting `.gitattributes`
  const parent = new GitHubProject({ name: "parent" });
  const child = new GitHubProject({
    parent,
    name: "child",
    outdir: path.join("packages", "child"),
  });

  // WHEN
  new TextFile(child, "file.txt", { lines: ["Content"] });

  // THEN
  const out = synthSnapshot(parent);
  expect(out[".gitattributes"].includes("file.txt")).toBe(false);
  expect(out["packages/child/.gitattributes"].includes("file.txt")).toBe(true);
});

// a project that depends on generated files during preSynthesize()
class PreSynthProject extends Project {
  public file: TextFile;
  public fileExistedDuringPresynth: boolean;
  constructor(options: Omit<ProjectOptions, "name"> = {}) {
    super({ name: "presynth-project", ...options });

    this.file = new TextFile(this, "presynth.txt", { lines: [PROJEN_MARKER] });
    this.fileExistedDuringPresynth = false;
  }

  preSynthesize() {
    this.fileExistedDuringPresynth = fs.existsSync(this.file.absolutePath);
  }
}
