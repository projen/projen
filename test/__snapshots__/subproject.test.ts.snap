// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`subprojects use root level default task 1`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "tasks": {
    "build": {
      "description": "Full release build",
      "name": "build",
      "steps": [
        {
          "spawn": "pre-compile",
        },
        {
          "spawn": "compile",
        },
        {
          "spawn": "post-compile",
        },
        {
          "spawn": "test",
        },
        {
          "spawn": "package",
        },
      ],
    },
    "compile": {
      "description": "Only compile",
      "name": "compile",
    },
    "default": {
      "description": "Synthesize project files",
      "name": "default",
      "steps": [
        {
          "cwd": "..",
          "exec": "npx projen default",
        },
      ],
    },
    "package": {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "pre-compile": {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": {
      "description": "Run tests",
      "name": "test",
    },
  },
}
`;

exports[`subprojects use root level default task 2`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".",
  "tasks": {
    "build": {
      "description": "Full release build",
      "name": "build",
      "steps": [
        {
          "spawn": "pre-compile",
        },
        {
          "spawn": "compile",
        },
        {
          "spawn": "post-compile",
        },
        {
          "spawn": "test",
        },
        {
          "spawn": "package",
        },
      ],
    },
    "compile": {
      "description": "Only compile",
      "name": "compile",
    },
    "default": {
      "description": "Synthesize project files",
      "name": "default",
      "steps": [
        {
          "cwd": "../..",
          "exec": "npx projen default",
        },
      ],
    },
    "package": {
      "description": "Creates the distribution package",
      "name": "package",
    },
    "post-compile": {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "pre-compile": {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "test": {
      "description": "Run tests",
      "name": "test",
    },
  },
}
`;
